### DISCUSSION TITLE
What new packages would you like to have on Termux

### DISCUSSION THREAD

**Discussion by thunder-coding** [termux/termux-packages#7199] [Ideas]
Termux has got 1300+ packages and is still growing. There may be many projects which are cool, useful as well as fun to have, and which we may have missed adding here. If you have got a awesome package you are in need of or something cool you created. Make sure you share it with the Termux users too

Heres the criteria for commenting and asking for a package here:
- No packages that just contain scripts
- No hacking/phishing packages
- Should be usefull
- The package should be Open Source
- Morever every package requests should follow [CONTRIBUTING.md](https://github.com/termux/termux-packages/blob/master/CONTRIBUTING.md#a-note-about-package-requests)

**Comment by ghost** [comment]
A fast Emulator

**Reply by ghost** [reply]
fast emulator is bochs or dosbox but limited to dos programs

**Reply by ghost** [reply]
Emulator and fast are mutually exclusive things.

**Comment by hnmn** [comment]
Deno 🦕
May be official vscode port🙄

**Reply by thunder-coding** [reply]
Yeah that might be a great addition to the x11 repository. But the hardest part of this will be Chromium and Electron port

**Reply by thunder-coding** [reply]
Also deno seems to run some pre-built binaries for building itself see [denoland/deno_third_party](https://github.com/denoland/deno_third_party)

**Reply by creepy-pasta101** [reply]
Well.. electron has problems with some custom command it tries to execute during its build process... And that too.. hardcoded...

I don't think chromium would be a challenge.... We may need gn to be added to `./scripts/setup-ubuntu.sh` and a few more changes to allow it to build.. I haven't tried it tho...

**Reply by tathastu871** [reply]
vscode available for proot-distro andronix

**Reply by tomap** [reply]
Deno => done :)

**Comment by hnmn** [comment]
Wait i forgot 
Openjdk will be great
I think having compilers is most important
Then installing or running other packages would be easy

**Comment by ghost** [comment]
We already have lots of pending package requests, see https://github.com/termux/termux-packages/issues?q=is%3Aopen+is%3Aissue+label%3A%22package+request%22. Would be interesting to know what exactly from this list you are consider useful and want to see packaged in Termux officially.

P.S. Is there someone who successfully compiled OpenJDK with at least Android NDK r21? It can be compiled with old NDK but the support of currently used one is a prerequisite of making OpenJDK available.

**Reply by hnmn** [reply]
I know one repo they manage to compile jdk 17 with ndk r21 
Take look at 
https://github.com/PojavLauncherTeam/android-openjdk-build-multiarch
On buildjre16 branch
Also artifacts are available i tried jdk 17
It run .jar files well but jshell and gradle other things have some issues

**Reply by ghost** [reply]
It uses NDK r10e which isn't suitable for termux-packages. I know about this project and I have specifically asked about **newer NDK**.
> It can be compiled with old NDK but the support of currently used one is a prerequisite of making OpenJDK available.

In fact, I can compile OpenJDK with r17 (last supporting `gcc` cross toolchain), but again our current NDK is r21 and is subject to be upgraded to even newer.

**Reply by hnmn** [reply]
https://github.com/PojavLauncherTeam/android-openjdk-build-multiarch/blob/f0ba0a3ece9adfff969a3df5062a03dfd2bad937/setdevkitpath.sh#L3

**Reply by hnmn** [reply]
![IMG_20210729_141517](https://user-images.githubusercontent.com/26086203/127461391-da087cc1-03b1-4970-bcc1-4bc6a4deb340.jpg)

https://github.com/PojavLauncherTeam/android-openjdk-build-multiarch/runs/2998382781?check_suite_focus=true#step:4:115

Looking at actions log they actually use ndk r21 but they extract standalone toolchain also 
They use gcc wrapper something i don't know

**Reply by hnmn** [reply]
![IMG_20210729_143802](https://user-images.githubusercontent.com/26086203/127464789-206960cf-53f0-44e9-9041-c11c1c0e1c6d.jpg)

**Reply by creepy-pasta101** [reply]
The Problem with openJDK currently is that it checks compiler sanity by checking its version.... Also currently openJDK supports cross compilation only with gcc

**Reply by ghost** [reply]
@suhan-paradkar It successfully compiles with NDK r21 using wrappers. I'm rewriting it as `build.sh` now, but the issue is `alsa` which is somehow fixed in the mentioned repo.

P.S. Using branch buildjre16 which succeed on CI.

**Reply by creepy-pasta101** [reply]
@xeffyr I saw you were working on openjdk17 with a lot of commits.. wouldn't it be better if patches were applied to @termux 's fork of openJDK instead of keeping it at `packages/openjdk-17` ?

**Reply by ghost** [reply]
This fork is temporary and is more like a backup since package is based on unofficial openjdk sources. Possibly that package will use the official openjdk repo in future.

Note that arm and i686 variants of openjdk are broken and this is most likely fault of https://github.com/PojavLauncherTeam/android-openjdk-build-multiarch/, as I can reproduce issues (https://github.com/termux/termux-packages/issues/7214, https://github.com/termux/termux-packages/issues/7215) with their builds too. 64-bit variants seems working fine.

**Reply by creepy-pasta101** [reply]
Well.. I guess openJDK's official mobile variant for openJDK17 can be used.. where they claim they support mobile devices [here](https://github.com/openjdk/mobile)

**Comment by SDRausty** [comment]
>What new packages would you like to have on Termux?

I would like to see `gradle` and https://github.com/skyleecm/android-build-tools-for-arm and the Android SDK packaged in Termux;  How about you?

**Reply by creepy-pasta101** [reply]
gradle is already there in the repo

**Reply by ghost** [reply]
> the Android SDK packaged in Termux

Just FYI: disk space on our package server is **finite** and is not so big (we have used more than 50% of disk space). Lots of small packages are better than one big suitable for very narrow purpose. Don't forget that besides main repo there are games, root, science, unstable and x11 repositories. They all share same disk space as main repo.

Check out https://github.com/termux/termux-packages/blob/master/CONTRIBUTING.md#packaging-policy which clearly states:
> Not taking too much disk space
>
>The size of resulting package should be less than 100 MiB.
>
>Since software is being compiled for 4 CPU architectures (aarch64, arm, i686, x86_64), effective disk usage is 4 times bigger than size of single .deb file. We are limited on disk space and prefer many small packages over one big.
>
>Exceptions are made on individual basis and only for packages providing important functionality.

**Reply by SDRausty** [reply]
> gradle is already there in the repo
```
$ gradle
No command gradle found, did you mean:
 Command grape in package groovy
```
Thank you for the quick replies:
`gradle -v`

`Welcome to Gradle 7.2-rc-1!`

**Comment by jeevatattva** [comment]
[`flite`](http://www.speech.cs.cmu.edu/flite/): a small, fast run time synthesis engine. Has an executable binary, along with a library, which can be integrated with `ffmpeg` as an external library. Currently, doesn't compile with `android-ndk` due to function overloading not mentioned explicitly.
[`frei0r`](https://frei0r.dyne.org/): minimalistic plugin API for video effects. Can be used with ffmpeg.

**Reply by thunder-coding** [reply]
The link for `flite` project you sent (`https://www.speech.cs.cmu.edu/flite/`) seems to be broken

**Reply by jeevatattva** [reply]
It's a `http` only site, and the URL opens for me now.

**Reply by tathastu871** [reply]
http://www.festvox.org/flite/

**Comment by sandeep-gh** [comment]
Python 3.10 would be great :).  Will soon try to build one myself. Got some pointers from Python folks.

**Comment by sandeep-gh** [comment]
I am using sshfs to share files between linux and the termux. Its bit flaky mostly because wifi on chromebook is very choppy.  Would be great some other file sharing tool that works over intermittent connections.

**Reply by ghost** [reply]
Try `croc`, however not sure about "intermittent connections".

**Comment by sandeep-gh** [comment]
croc wont fit the bill, however seems like nice tool to know about. I want to expose termux files as part of filesystem on remote machine for which sshfs or samba service is suited. I tried sshfs but its flaky. Not sure if its possible to build samba on termux.

**Comment by araczkowski** [comment]
USB serial without serial module in kernel is in my opinion the last puzzle to solve and have all functionality

Maybe this way is possible?
https://github.com/jacklinquan/usbserial4a

**Comment by utmost-premise** [comment]
> No packages that just contain scripts

Why not? I would like to have [`tudo`](https://github.com/agnostic-apollo/tudo), so it could be installed and updated through `pkg`

**Reply by ghost** [reply]
Termux official repository is not a place for collecting random stuff from the whole Internet. https://github.com/termux/termux-packages/blob/master/CONTRIBUTING.md#packaging-policy

Scripts doesn't require compilation, just download from repo and use it. Why need package?

**Reply by utmost-premise** [reply]
> Why need package?

So I don't have to remember manually looking for updates.

But thank you anyway.

**Comment by zxq432** [comment]
gentoo's portage package manager. so that all termux-packages could be easily merged/updated from source code to termux directly from termux without cross compile on PC first.  it's not a single package, since it depend on other packages. gentoo is already a proot-distro release, and current mobile phone is powerful enough to do the compiling work. with the portage package manager, the C flags could be optimized for a specific soc to improve performance. for example CFLAGS="-march=armv8.2-a+crypto -mcpu=cortex-a76.cortex-a55+crypto -mtune=cortex-a76.cortex-a55". and the dependecy is well defined by runtime dependency, build time dependency and feature dependencies. fine ebuild phase with hook support, build.log wiht debug support for each package merged, slot for multiple version installation of some pacakges, such as install python3.{7,8,9,10}, while select python3.9 currently system wide phthon. gentoo portage surport prefix and gentoo prefix could co-exist smoothly with other GNU/linux.

**Comment by sandeep-gh** [comment]
This looks certainly interesting, if feasible. If I understand correctly,
Instead of porting a package by package, if we somehow make existing
package manager that would be great. I just don't know the internals to
know if this is feasible especially for more close to OS system software --
postgres, samba, etc.

**Reply by ghost** [reply]
You're looking for proot'ed distribution. There no way to turn Termux environment into traditional distribution. Moreover, that's not it's goal. Termux is extended Linux user space environment for Android OS. If that chosen Termux road bothers you, please use alternatives such as https://github.com/CypherpunkArmory/UserLAnd or Linux Deploy.

postgres is available as `postgresql` package.

`samba` is present in unstable repo.

**Reply by sandeep-gh** [reply]
Just checked out PRooted distrubution. Fits the bill perfectly. THanks.

**Comment by sandeep-gh** [comment]
Currently, termux sits well perfectly in my workflow. I have some
requirements for other packages/enviorments (including samba) that I am
exploring. Was interested in that regard.
 Will look into the alternatives as well. Thanks for those too.

**Reply by leechaowen** [reply]
r-base

**Comment by agnostic-apollo** [comment]
@leechaowen `r-base` is provided by `its-pointless` repo if you don't know already.

https://github.com/its-pointless/gcc_termux

https://wiki.termux.com/wiki/Package_Management#Community_repositories

**Reply by leechaowen** [reply]
libxml2-dev

**Reply by ghost** [reply]
you can install that via `libxml2` package

**Reply by leechaowen** [reply]
yes . i know  its-pointless.  but once i run pkg update & pkg upgrade in termux , the R could not workding. it seems such action broke the dependies.  

I'm on learning how to compile r-base source code in termux recently. 

~ $ R
CANNOT LINK EXECUTABLE "/data/data/com.termux/files/usr/lib/R/bin/exec/R": library "libicuuc.so.69" not found: needed by main executable
